internal/node/monitor/monitor.go

```diff
@@ -98,18 +98,33 @@ func (m *Monitor) getWorkerIndexingStateByClients(ctx context.Context, n network
 	return client.CurrentState(state), latestState, nil
 }
 
-// flagUnhealthyWorker detects by comparing the current and latest block height/number. If the difference is greater than the tolerance, the worker is flagged as unhealthy.
-func (m *Monitor) flagUnhealthyWorker(ctx context.Context, network, workerID, workerName string, currentWorkerState, latestWorkerState, networkTolerance uint64) error {
-	currentStatus := m.GetWorkerStatus(ctx, network, workerName)
+// flagUnhealthyWorkerByID detects by comparing the current and latest block height/number. If the difference is greater than the tolerance, the worker is flagged as unhealthy.
+func (m *Monitor) flagUnhealthyWorkerByID(ctx context.Context, workerID string, currentWorkerState, latestWorkerState, networkTolerance uint64) error {
+	currentStatus := m.GetWorkerStatusByID(ctx, workerID)
+	targetStatus := currentStatus
 
 	if latestWorkerState < currentWorkerState {
 		return nil
 	}
 
-	if (currentStatus == workerx.StatusReady && latestWorkerState-currentWorkerState > networkTolerance) || (currentStatus == workerx.StatusIndexing && currentWorkerState <= m.getWorkerProgress(ctx, workerID)) {
-		if err := m.UpdateWorkerStatus(ctx, network, workerName, workerx.StatusUnhealthy.String()); err != nil {
-			return fmt.Errorf("update worker status: %w", err)
+	switch currentStatus {
+	case workerx.StatusReady:
+		// if the worker is ready but the difference between the current and latest block height/number is greater than the tolerance, flag it as unhealthy
+		if latestWorkerState-currentWorkerState > networkTolerance {
+			targetStatus = workerx.StatusUnhealthy
```

Comment:
if the block lag is detected while in StatusReady, it would be more reasonable to switch to StatusIndexing, and then check StatusIndexing again next time. 🤔